pkg/controllers/networkinfo/networkinfo_controller.go

```diff
@@ -297,45 +269,152 @@ func (r *NetworkInfoReconciler) Start(mgr ctrl.Manager) error {
+	return nsSet, idSet, nil
+}
+
+// CollectGarbage logic for NSX VPC is that:
 // 1. list all current existing namespace in kubernetes
-// 2. list all the nsx-vpc in vpcStore
-// 3. loop all the nsx-vpc to get its namespace, check if the namespace still exist
-// 4. if ns do not exist anymore, delete the nsx-vpc resource
+// 2. list all the NSX VPC in vpcStore
+// 3. loop all the NSX VPC to get its namespace, check if the namespace still exist
+// 4. if ns do not exist anymore, delete the NSX VPC resource
 // it implements the interface GarbageCollector method.
 func (r *NetworkInfoReconciler) CollectGarbage(ctx context.Context) {
-	log.Info("VPC garbage collector started")
-	// read all nsx-vpc from vpc store
+	startTime := time.Now()
+	defer func() {
+		log.Info("VPC garbage collection completed", "duration(ms)", time.Since(startTime).Milliseconds())
+	}()
+
+	// read all NSX VPC from VPC store
 	nsxVPCList := r.Service.ListVPC()
 	if len(nsxVPCList) == 0 {
+		log.Info("No NSX VPCs found in the store, skipping garbage collection")
 		return
 	}
 
-	// read all namespaces from k8s
-	namespaces := &corev1.NamespaceList{}
-	err := r.Client.List(ctx, namespaces)
+	_, idSet, err := r.listNamespaceCRsNameIDSet(ctx)
 	if err != nil {
-		log.Error(err, "failed to list k8s namespaces")
+		log.Error(err, "Failed to list Kubernetes Namespaces for VPC garbage collection")
 		return
 	}
-	nsSet := sets.NewString()
-	for _, ns := range namespaces.Items {
-		nsSet.Insert(ns.Name)
-	}
 
-	for i := len(nsxVPCList) - 1; i >= 0; i-- {
-		nsxVPCNamespace := getNamespaceFromNSXVPC(&nsxVPCList[i])
-		if nsSet.Has(nsxVPCNamespace) {
+	for i, nsxVPC := range nsxVPCList {
+		nsxVPCNamespaceName := filterTagFromNSXVPC(&nsxVPCList[i], commonservice.TagScopeNamespace)
+		nsxVPCNamespaceID := filterTagFromNSXVPC(&nsxVPCList[i], commonservice.TagScopeNamespaceUID)
+		if idSet.Has(nsxVPCNamespaceID) {
 			continue
 		}
-		elem := nsxVPCList[i]
-		log.Info("GC collected nsx VPC object", "ID", elem.Id, "Namespace", nsxVPCNamespace)
+		log.Info("Garbage collecting NSX VPC object", "VPC", nsxVPC.Id, "Namespace", nsxVPCNamespaceName)
 		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, common.MetricResTypeNetworkInfo)
-		err = r.Service.DeleteVPC(*elem.Path)
-		if err != nil {
+
+		if err = r.Service.DeleteVPC(*nsxVPC.Path); err != nil {
+			log.Error(err, "Failed to delete NSX VPC", "VPC", nsxVPC.Id, "Namespace", nsxVPCNamespaceName)
 			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, common.MetricResTypeNetworkInfo)
-		} else {
-			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, common.MetricResTypeNetworkInfo)
+			continue
 		}
+
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, common.MetricResTypeNetworkInfo)
+		log.Info("Successfully deleted NSX VPC", "VPC", nsxVPC.Id)
 	}
 }
+
+func (r *NetworkInfoReconciler) deleteVPCsByName(ctx context.Context, ns string) error {
```

Comment:
deleteVPCsByName and deleteVPCsByID have much in common, it's better to refactor them.

Reply:
done.